mlir/lib/Dialect/Linalg/Transforms/HoistVectorTransfers.cpp

```diff
@@ -0,0 +1,267 @@
+
+    if (reductionCount == 1)
+      return rewriter.notifyMatchFailure(
+          contractOp, "Batch matmul operation not supported yet");
+
+    if (reductionCount > 2)
+      return rewriter.notifyMatchFailure(
+          contractOp, "The vector contract operation is not a gemm");
+
+    // Check the K-dim to be 1
+    int64_t K =
+        vectorReadOpLhsType.getDimSize(vectorReadOpLhsType.getRank() - 1);
+    if (K != 1)
+      return rewriter.notifyMatchFailure(contractOp, "K dim is not 1");
+
+    // Check whether the BR-matmul tiling + vector contract pattern matches for the
+    // 4-nested loop structure
+    auto loops = getNestedLoop(contractOp);
+    if (failed(loops))
+      return rewriter.notifyMatchFailure(
+          contractOp, "Invalid loop nest in contract pattern");
+
+    auto checkLoops = checkNestedLoop(*loops, *subviews);
+    if (failed(checkLoops))
+      return rewriter.notifyMatchFailure(
+          contractOp, "Loops doesn't match the iv in subviews");
+
+    auto nestedLoops = *loops;
+    auto kForOp = nestedLoops[0];
+    auto reductionForOp = nestedLoops[1];
+
+    // Move the vector transfer read before the reduction and k loop
+    rewriter.setInsertionPoint(reductionForOp);
+    auto *cloneVectorReadOp = rewriter.clone(*vectorReadOpAcc);
+
+    // Code to re-create the reduction and k loop with iter args
+    auto vectorReadOpValue = cloneVectorReadOp->getResult(0);
+    auto newReductionForOp = rewriter.create<scf::ForOp>(
+        reductionForOp.getLoc(), reductionForOp.getLowerBound(),
+        reductionForOp.getUpperBound(), reductionForOp.getStep(),
+        ValueRange{vectorReadOpValue},
+        [&](OpBuilder &rewriterNewReductionForOp, Location locNewReductionForOp,
+            Value ivNewReductionForOp, ValueRange iterArgsNewReductionForOp) {
+          auto newKForOp = rewriter.create<scf::ForOp>(
+              kForOp.getLoc(), kForOp.getLowerBound(), kForOp.getUpperBound(),
+              kForOp.getStep(), iterArgsNewReductionForOp,
+              [&](OpBuilder &rewriterNewKForOp, Location locNewKForOp,
+                  Value ivNewKForOp, ValueRange iterArgsNewKForOp) {
+                IRMapping mapper;
+                mapper.map(reductionForOp.getInductionVar(),
+                           ivNewReductionForOp);
+                mapper.map(kForOp.getInductionVar(), ivNewKForOp);
+
+                for (auto &op : kForOp.getBody()->without_terminator()) {
+                  rewriterNewKForOp.clone(op, mapper);
+                }
+                rewriterNewKForOp.create<scf::YieldOp>(locNewKForOp,
+                                                       iterArgsNewKForOp);
+              });
+          rewriterNewReductionForOp.create<scf::YieldOp>(
+              locNewReductionForOp, newKForOp.getResult(0));
+        });
+
+    // Code to hoist vector transfer write after reduction loop and also to
+    // update the yield of k loop
+    auto newKForOp =
```

Comment:
Here as well: I am not sure why this section appears a bit involved. Per the example in the docs, you are just hoisting one transfer_read and "threading" that vector through the iter_args of two loops, right? Doesn't two applications of `LoopLikeOpInterface::replaceWithAdditionalYields` suffice for this?